Create threadpool for GC parallel tasks

schveiguy opened the pull request:
When this is merged, we will replace the parallel scan with this thread pool.

Next step will be to start the threads before stopping the world to prevent deadlocks.

Then we can work on some parallel collect features.

schveiguy committed:
[GC] Add threadpool module, replace parallel scanning thread
infrastructure with threadpool. Also adds mechanism for thread to detach
itself (threadpool worker threads should not be paused for scanning)

schveiguy committed:
[GC] Add threadpool unittests

schveiguy committed:
[GC] Add some threadpool comments.

deadalnix reviewed (changes requested):
IMO, we should focus on the thread pool itself before changing bits and pieces all over the place.

It is not clear at this time that most of them are required. It is also unclear to me that the current API makes a lot of sense, but that would be easier to discuss in a context where that doesn't impact integration.

For instance, shouldn't the main thread be put to use when waiting, for instance when joining or waiting for idle? I think this is the natural place to do so.

schveiguy commented:
> shouldn't the main thread be put to use when waiting

The main thread is put to work if you pass `true` to `useThisThread`.

schveiguy commented:
Oh I see, instead of doing that, you do it inside the waiting functions.

I think I can do this by moving the boolean to those functions. I think you should still be able to ensure the main thread does not run the delegate (it may do things that are incompatible with the main thread).

schveiguy commented:
FWIW, integrating discovered a lot of problems with the threadpool that I likely would not have found in artificial unittests. So while I can split out the integration, I'm glad I did it before proposing this PR.

schveiguy committed:
[GC] Remove threadpool from scanner use for now.

schveiguy committed:
[GC] Rework threadpool API so the calling thread is only used when
waiting for completion or joining of threads, simplifies the dispatch
API. Combine the two execution functions to avoid so much code
duplication.

schveiguy commented on sdlib/d/gc/threadpool.d:
```diff
@@ -0,0 +1,273 @@
+		}
+	}
+
+	void dispatchImpl(void delegate(uint) work, uint count) {
+		assert(mutex.isHeld(), "Mutex not held!");
+		auto sharedThis = cast(shared(ThreadPool)*) &this;
+		sharedThis.mutex.waitFor(noMoreWork);
```
This was puzzling for me. `Mutex.waitFor` requires a shared mutex, yet the mutex must be locked to call it. Typically, the mutex is unshared in a non-shared method, so this was awkward. What is the reason for `waitFor` being shared-only?

You see an interesting thing here as well: https://github.com/snazzy-d/sdc/blob/f961950b1dce0026b11ecdc7ac0ebb386486572d/sdlib/d/gc/scanner.d#L173

There, you are in a shared method, but you need to call an unshared method because the thing is locked at the time.

I suppose I could mark the mutex itself shared, so it's always compatible?

schveiguy commented on sdlib/d/gc/threadpool.d:
```diff
@@ -0,0 +1,273 @@
+	 * Wait for all work to be complete. `useThisThread` set to true will
+	 * use the calling thread to also do work if needed.
+	 */
+	void waitForIdle(bool useThisThread = true) shared {
```
> First there is a bit of featuritis, as why wouldn't we want to get more work done faster?

So in the scanner, there is a difference in running the scan function from the main thread, and running it from the worker threads. Namely, in the worker threads, the GC is turned off, whereas in the main thread it is not turned off. Although I moved this code to the worker delegate anyway, since turning off the GC is harmless at this point (it has already been turned off), it got me thinking, you might want to do some things in the worker threads that would be incorrect for the main thread. So I added this flag as an option.

I can remove this feature for now, and we can add it when we need it.

> But second, we assume no more work will be added once we exhaust the current stash of work, but we have no way to know that, and if that happens, we'll be stuck waiting again instead of using this thread.

Work will not be added because the main thread is the only one who can add work. If you tried to dispatch more tasks while in one of the worker threads, I think there could be a race condition.

Note that this mechanism is not necessarily how the threads will divide the actual work (e.g. the scanning of a specific range). This abstraction is for creating threads, and having them run a specific task. Anything involving a changing list of work should be handled externally. There is no allocating inside here. The goal of this is to manage threads specifically for the GC, which is a restrained environment.

schveiguy replied:
> I'm sure there is a ton of literature on how to build a thread pool out there that don't have that limitation, no?

I would expect building a thread pool which is heavily constrained to avoid allocations and other requirements is not very well documented. I can look though.

> it is pretty obvious that at some point we'll want to execute heterogeneous workload

This is supported. e.g. you have 10 worker threads, you can dispatch 5 of them to do task 1, and then 5 of them to do task 2. The way this works is that the call waits for all 5 tasks to be dispatched, and then dispatches the 5 new tasks. It does not wait for all threads to be idle before dispatching the second, so you should in effect get all 10 threads running in parallel.

I did it this way to avoid allocating a work list of things to do (we can't do allocations here).

schveiguy commented on sdlib/d/gc/tstate.d:
```diff
@@ -76,6 +76,17 @@ public:
+
+		// should only ever be called while in a normal thread state.
+		size_t s = RunningState;
+		auto succeeded = state.casWeak(s, SuspendState.Detached);
+		assert(succeeded);
```
> First, you can't ensure cas won't fail, but second, you can't ensure the thread is in a running state as a signal can be sent at anytime.

No, this can't happen. If it got a signal, it would not be able to detach itself, it would be in the signal handler. Note from the assert above, only a thread can detach itself explicitly, it can't be done outside the thread itself.

Besides, there is an assert here, the point is to ensure you only call this while in the right state. I want the bomb to explode if you don't do it right.

I suppose I could call store, and then validate the state was running?

> This is also impossible to review in its form. The behavior of all detach in the codebase are now changed for this, but we don't know where they are.

I grepped, there is one spot. Any currently valid code would only call detach while in the `Signaled` state. and I changed that one spot already.

I'll try and fix the names so I don't have to change the existing one. I felt like `detach` was the correct name for this.

schveiguy replied:
I can make it a loop.

schveiguy commented on sdlib/d/gc/tstate.d:
```diff
@@ -62,7 +62,7 @@ public:
 	}
 
-	void detach() {
+	void assumeDetached() {
```
We are assuming the thread isn't going to respond to signals. We are assuming it detached by blocking signals.

schveiguy commented on sdlib/d/gc/threadpool.d:
```diff
@@ -0,0 +1,273 @@
+		sum.fetchAdd(idx);
+	}
+
+	threadPool.dispatch(otherTask, 128);
```
ok

schveiguy commented on sdlib/d/gc/threadpool.d:
```diff
@@ -0,0 +1,273 @@
+
+unittest threadpool {
+	import core.stdc.pthread;
+	import core.stdc.unistd;
```
I normally would write `import core.stdc.unistd : sleep`, but that's not supported.

deadalnix reviewed (changes requested):
Sorry to have dropped that ball.

There is definitively a problem with the CAS, for the thread pool, I'm not convinced this design is not a problem in the making. We should at least pretend we did it right in its API, even if it won't work with this implementation.
